codec-http2/src/main/java/io/netty/handler/codec/http2/Http2FrameCodec.java

```diff
@@ -145,6 +145,7 @@
     private static final InternalLogger LOG = InternalLoggerFactory.getInstance(Http2FrameCodec.class);
 
     private final PropertyKey streamKey;
+    private final PropertyKey upgradeKey;
```

Comment:
can you add some javadocs indicating why this `PropertyKey` is necessary? (/cc @ejona86 https://github.com/netty/netty/issues/7280#issuecomment-334593812). If we are just trying to prevent the server from making contributions to flow control because the stream is created in "half closed local" on the client [1] then can we just do: 

```java
if (streamId == HTTP_UPGRADE_STREAM_ID && connection().isServer()) {
  return false;
}
...
```

Should be done in a lower layer (e.g. `DefaultHttp2LocalFlowController#consumeBytes`)?

[1] https://tools.ietf.org/html/rfc7540#section-5.1.1
> HTTP/1.1 requests that are upgraded to HTTP/2 (see Section 3.2) are
   responded to with a stream identifier of one (0x1).  After the
   upgrade completes, stream 0x1 is "half-closed (local)" to the client.
   Therefore, stream 0x1 cannot be selected as a new stream identifier
   by a client that upgrades from HTTP/1.1.

Reply:
@Scottmitch I think h2 clients use 0x1 as the first stream id–I think that designation is only true for h2c clients.

Reply:
Good call.

Any reason why we cannot move this `upgradeKey` into lower levels (e.g. when the stream is created in `Http2ConnectionHandler` and in `DefaultHttp2LocalFlowController#consumeBytes`)? It seems like this should always happen (even if not using the frame codec).

Reply:
Sure, I can try moving it.

Reply:
@Scottmitch hmm, I actually don't know how to thread the upgradeKey through.  any ideas?

Reply:
Good question. This isn't exactly straightforward, and seems like we may need to make some sort of API change. Here are some options:

1. Let the `Http2LocalFlowController` create the upgrade stream. We would need some sort of API such as `Http2LocalFlowController#createUpgradeStream(...)` and the `Http2LocalFlowController` could then keep the key local.
2. Add an API on the `Http2LocalFlowController` to notify that a remote initiated upgrade stream was created.
3. Add a method on the `Http2Stream` interface to determine if the stream was a remote initiated update. The `Http2LocalFlowController` could use this method to determine if consumeBytes behavior. This state can be set at the `Http2ConnectionHandler` level, potentially with an additional method/overload for `Http2Connection#createStream`. 

Option 1 leaks some responsibilities between layers, and we would have to be careful the correct endpoint and initial stream state is used.
Option 2 increases the API footprint of `Http2LocalFlowController`, but doesn't leave much room for error and keeps concerns separated.
Option 3 will require additional APIs and also possibly additional state for every stream (e.g. a boolean) that *may* only be used for a single stream.

If there are no other options my preference would be for option (2).

Reply:
I also think 2) would be good.

Reply:
@ejona86 - any better ideas?

Reply:
also @carl-mastrangelo @nmittler

Reply:
I know I'm late in replying... I don't have any better ideas. I agree that this should be done at a lower level, but it wasn't clear to me how we'd want to do that (which is what options you discussed). I would have needed to dig in much more than I have time to do. So I figured workaround here as a first-go, and we either fix it up later or just accept it and bide our time until the old API goes away.

Reply:
I am ok with this... @Scottmitch WDYT ?